File: code/ch11/11.2.1_minwindow.py
class Solution:
    def minWindow(self, s:str, t:str)->str:
        if not t or not s:
            return ""
            
        dict_t = Counter(t)
        
        required = len(dict_t)
        
        #初始化滑动窗口
        left, right = 0, 0
        
        formed = 0
        
        window_counts = {}
        
        ans = float("inf"), None, None
        
        while right < len(s):
            character = s[right]
            window_counts[character] = window_counts.get(character, 0) + 1
            
            if character in dict_t and window_counts[character] == dict_t[character]:
                formed += 1
                
            while left <= right and formed == required:
                character = s[left]
                
                if right - left + 1 < ans[0]:
                    ans = (right - left + 1, left, right)
                    
                window_counts[character] -= 1
                if character in dict_t and window_counts[character] < dict_t[character]:
                    formed -= 1
                    
                left += 1
            right += 1
                
        if ans[0] == float("inf"):
            return ""
        else:
            return s[ans[1]: ans[2] + 1]

File: code/ch11/11.2.1_minwindow.java
class Solution {
    public String minWindow(String s, String t) {
        if (s.length() == 0 || t.length() == 0)
            return "";
        int left = 0, right = 0;
        Map<Character, Integer> tMap = new HashMap<>();
        for (char c : t.toCharArray()) {
            tMap.put(c, tMap.getOrDefault(c, 0) + 1);
        }

        int required = tMap.keySet().size(), formed = 0;
        Map<Character, Integer> windowMap = new HashMap<>();

        // ans[0] 存放字符长度，ans[1]和ans[2]分别存放左右索引
        int[] ans = new int[3];
        ans[0] = Integer.MAX_VALUE;

        while (right < s.length()) {
            char character = s.charAt(right);
            windowMap.put(character, windowMap.getOrDefault(character, 0) + 1);

            if (tMap.containsKey(character) && windowMap.get(character).intValue() == tMap.get(character).intValue())
                formed++;

            while (left <= right && (formed == required)) {
                character = s.charAt(left);

                if (right - left + 1 < ans[0]) {
                    ans = new int[]{right - left + 1, left, right};
                }
                windowMap.put(character, windowMap.getOrDefault(character, 1) - 1);

                if (tMap.containsKey(character) && windowMap.get(character) < tMap.get(character))
                    formed -= 1;

                left++;
            }
            right++;
        }

        return ans[0] == Integer.MAX_VALUE ? "" : s.substring(ans[1], ans[2] + 1);
    }
}